### ISSUE TITLE
How to handle dynamic addition/removal of sprites?

### ISSUE THREAD

**Issue by adroitwhiz** [PullJosh/leopard#68]
There's some broader discussion to be had about the ability to mess around with a project's `sprites` after creating the project, and the best way to expose that. I talked about this a bit [in a comment on another issue](https://github.com/PullJosh/leopard/issues/34#issuecomment-570013424), but this is important because it touches the API surface and answers a lot of currently-open questions about various things in the codebase (e.g. loading projects and assets, layer ordering).

Right now, you can change which sprites appear in a project by adding to the project's `sprites` object.

This means that if Leopard wants to run some of its own code when you add or remove a sprite (e.g. to add it to / remove it from an internal list of sprites in order to properly do layer ordering), it can't do that under the current API (in which you just add properties to the `project.sprites` object) unless we decide to "fake" `project.sprites` being an object by using a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) (which, as I mentioned earlier, cannot be polyfilled unlike the other ES6 features we use in this codebase, and seems like a janky way to implement things anyhow).

Scratch has its own internal model for how sprites/clones work: a "sprite" is the "prototype" object that holds the blocks, costumes, and sounds, and a "target" is a specific instance of a prototype "sprite" which holds things like transformation, current costume, and other state. The thing that differs from other languages is that you also always have a single instance of the sprite "prototype" on the stage, and clones are created from that.

We could require all sprite "prototypes" to be declared upfront (like in Scratch, where sprites can't create entire new sprites, only clones that share their own code/costumes/sounds); when you construct a `Project`, you pass in all the `Sprite`s that you will use in it, and then it will load all the costumes and sounds, let the renderer create WebGL textures, and do all the other prep work that needs to be done before starting the project. You can then only create additional sprites by cloning them.

We could also add a specific way to add or remove sprites-- a `Project.addSprite` function, for example, could return a `Promise` that resolves once Leopard is done doing things like loading the sprite's costumes. We could expose a similar API for dynamically adding costumes/sounds, if desired (currently, costumes just won't show up and sounds just won't play until they're loaded, which is weird).

There are likely other options; I just can't think of them at the moment. Right now, however, the API surface is super weird, blocks implementation of important features like layer ordering, and has "footgun" potential around the way assets (costumes + sounds) are loaded. It's important that we iron this out before too many Leopard projects end up relying on being able to modify a project's `sprites` object.

**Comment by PullJosh** [comment]
You're absolutely right. We need to talk about this, as well as layering (z-ordering of sprites), before too many Leopard projects are created.

**Comment by PullJosh** [comment]
One weird side-effect of dynamically loading sprites is that it will take time for assets (costumes/sounds) to be loaded, which can lead to a janky experience.

We currently take a pretty laissez faire approach to assets (just using them as they arrive, without any sort of loading screen up front). I think in the long run, a proper loading screen is probably a worthwhile addition.

However, if we allow adding new sprites dynamically, this approach breaks down. We would have to force users to register sprites up front but only add instances later.

**Comment by PullJosh** [comment]
Personally, I don't have a huge issue with mimicking Scratch's approach, where you have to have one base instance of a sprite but can create clones. This would be an easy, mostly backwards-compatible change, too, where we just freeze the sprite object.
